File: solutions/typescript/anagram/2/anagram.ts
export class Anagram {
    private lowerWord: string;
    private normalizedWord: string;

    constructor(word: string) {
        this.lowerWord = word.toLowerCase();
        this.normalizedWord = this.normalize(this.lowerWord);
    }

    private normalize(str: string): string {
        return str.split('').sort().join('');
    }

    public matches(...candidates: string[]): string[] {
        const result: string[] = [];
        const flatCandidates = candidates.flat();

        for (const candidate of flatCandidates) {
            const lowerCandidate = candidate.toLowerCase();
            if (this.lowerWord === lowerCandidate) {
                continue;
            }

            const normalizedCandidate = this.normalize(lowerCandidate);

            if (this.normalizedWord === normalizedCandidate) {
                result.push(candidate);
            }
        }

        return result;
    }
}

File: solutions/go/anagram/2/anagram.go
package anagram

import (
	"slices"
	"strings"
)

func normalize(s string) string {
	runes := []rune(s)
	slices.Sort(runes)
	return string(runes)
}

func Detect(subject string, candidates []string) []string {
	result := []string{}
	lowerSubject := strings.ToLower(subject)
	normalizedSubject := normalize(lowerSubject)

	for _, candidate := range candidates {
		if lowerSubject == strings.ToLower(candidate) {
			continue
		}
		if normalizedSubject == normalize(strings.ToLower(candidate)) {
			result = append(result, candidate)
		}
	}

	return result
}
